Google.Api.Gax.Grpc/Rest/RestChannel.cs

```diff
@@ -71,7 +71,7 @@ internal sealed class RestChannel : ChannelBase
 
             var httpResponseTask = cancellationContext.RunAsync(
                 cancellationToken => SendAsync(restMethod, host, options, request, HttpCompletionOption.ResponseContentRead, cancellationToken));
-            var readResponseTask = ReadResponseAsync(httpResponseTask);
+            var readResponseTask = ReadResponseAsync(httpResponseTask, cancellationContext);
```

Comment:
Ah, I see, you are disposing of that on ReadResponseAsync to dispose of it sooner and more reliably.

So that ReadResponseAsync does not have to recieve the cancellationContext (which is passed as null in streaming) and given that SendAsync is called with `ResponseContentRead`, would you consider instead disposing of it like this:

```
            var httpResponseTask = cancellationContext.RunAsync(cancellationToken =>
            {
                try
                {
                    return await SendAsync(restMethod, host, options, request, HttpCompletionOption.ResponseContentRead, cancellationToken);
                }
                finally
                {
                    cancellationSource.Dispose();
                }
            });
```

Reply:
Also, I think the refactor I proposed in the previous PR will help clearing this a lot. If you pass the new StreamingHttpResponseMessage (instead of the TextReader task) to the PartialDecoding.... then on the first MoveNext we can check for Content and if there is Content there was an error and we know to throw the RPCException there, else use the StreamingContent task (which would be the TextReader) normally.
If I'm not explaining myself well, I'm happy to give it a go.

Reply:
Will have a look at the first comment above when I'm at my laptop, but my initial instinct is to merge and then refactor later (possibly with the bigger refactoring you were describing).